core/src/main/java/org/kohsuke/stapler/RequestImpl.java

```diff
@@ -698,20 +698,35 @@ public Object convertJSON(Object o) {
                 if(l==null) {// single value conversion
                     try {
                         Class actualType = type;
+                        boolean isArray = false;
                         String className = null;
                         if(j.has("stapler-class")) {
+                            if (j.optJSONArray("stapler-class") != null) {
+                                isArray = true;
+                            }
                             // deprecated as of 2.4-jenkins-4 but left here for a while until we are sure nobody uses this
                             className = j.getString("stapler-class");
                             LOGGER.log(FINE, "stapler-class is deprecated in favor of $class: {0}", className);
                         }
                         if(j.has("$class")) {
+                            if (j.optJSONArray("$class") != null) {
+                                isArray = true;
+                            }
                             className = j.getString("$class");
                         }
 
                         if (className != null) {
                             // sub-type is specified in JSON.
                             // note that this can come from malicious clients, so we need to make sure we don't have security issues.
 
+                            if (isArray) {
+                                throw new IllegalArgumentException(
+                                        "The frontend sent an unexpected list of classes ("
+                                                + className
```

Comment:
What will an example value be?

Reply:
Please read [JENKINS-67550](https://issues.jenkins.io/browse/JENKINS-67550) (which I linked to in the PR description). It contains an example value seen in the wild:

> `["org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition","org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition"]`

Reply:
So we could also check, say,

```java
className.contains(",")
```
?

Reply:
Correct